Audio/vision task guides (#15808)
huggingface/transformers: docs/source/tasks/asr.mdx

<!--Copyright 2022 The HuggingFace Team. All rights reserved.

Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
the License. You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.
-->

# Automatic speech recognition

<Youtube id="TksaY_FDgnk"/>

Automatic speech recognition (ASR) converts a speech signal to text. It is an example of a sequence-to-sequence task, going from a sequence of audio inputs to textual outputs. Voice assistants like Siri and Alexa utilize ASR models to assist users.

> This guide will show you how to fine-tune [Wav2Vec2](https://huggingface.co/facebook/wav2vec2-base) on the [TIMIT](https://huggingface.co/datasets/timit_asr) dataset to transcribe audio to text.

Reviewer: I also think TIMIT is the best dataset to demo here. Just FYI, we might have to make it a manual dataset soon because of licensing cc @lhoestq . But I think then we can still provide provide a simple `wget ...` link

... )
```

Take a look at the example again:

```py
>>> timit["train"][0]
{'audio': {'array': array([-2.1362305e-04,  6.1035156e-05,  3.0517578e-05, ...,
         -3.0517578e-05, -9.1552734e-05, -6.1035156e-05], dtype=float32),
  'path': '/root/.cache/huggingface/datasets/downloads/extracted/404950a46da14eac65eb4e2a8317b1372fb3971d980d91d5d5b221275b1fd7e0/data/TRAIN/DR4/MMDM0/SI681.WAV',
  'sampling_rate': 16000},
 'file': '/root/.cache/huggingface/datasets/downloads/extracted/404950a46da14eac65eb4e2a8317b1372fb3971d980d91d5d5b221275b1fd7e0/data/TRAIN/DR4/MMDM0/SI681.WAV',
 'text': 'Would such an act of refusal be useful?'}
```

The `audio` column contains a 1-dimensional `array` of the speech signal that must be called to load and resample the audio file.

## Preprocess

Load the Wav2Vec2 processor to process the audio signal and transcribed text:

```py
>>> from transformers import AutoProcessor

>>> processor = AutoProcessor.from_pretrained("facebook/wav2vec2-base")
```

The preprocessing function needs to:

1. Call the `audio` column to load and resample the audio file.
2. Extract the `input_values` from the audio file.

> 2. Typically, when you call the processor, you call the feature extractor. Since you also want to tokenize text, instruct the processor to call the tokenizer instead with a context manager.

Reviewer 2: ```suggestion
3. Typically, when you call the processor, you call the feature extractor. Since you also want to tokenize text, instruct the processor to call the tokenizer instead with a context manager.
```

```py
>>> def prepare_dataset(batch):
...     audio = batch["audio"]
    

> ...     batch["input_values"] = processor(audio["array"], sampling_rate=16000).input_values[0]

Reviewer: ```suggestion
...     batch["input_values"] = processor(audio["array"], sampling_rate=audio["sampling_rate"]).input_values[0]
much safer to verify that the audio is correctly sampled

...     batch["input_length"] = len(batch["input_values"])

...     with processor.as_target_processor():
...          batch["labels"] = processor(batch["text"]).input_ids
...     return batch
```

> Use 🤗 Datasets [`map`](https://huggingface.co/docs/datasets/package_reference/main_classes.html#datasets.Dataset.map) function to apply the preprocessing function over the entire dataset. You can speed up the map function by setting `batched=True` to process multiple elements of the dataset at once and increasing the number of processes with `num_proc`. Remove the columns you don't need:

Reviewer: ```suggestion
Use 🤗 Datasets [`map`](https://huggingface.co/docs/datasets/package_reference/main_classes.html#datasets.Dataset.map) function to apply the preprocessing function over the entire dataset. You can increasing the number of processes with `num_proc`. Remove the columns you don't need:
```

This doesn't work well because batched audio loading / sampling is done in a loop currently under the hood no @lhoestq ?

```py
>>> timit = timit.map(prepare_dataset, remove_columns=timit.column_names["train"], num_proc=4)
```

🤗 Transformers doesn't have a data collator for automatic speech recognition, so you will need to create one. You can adapt the [`DataCollatorWithPadding`] to create a batch of examples for automatic speech recognition. It will also dynamically pad your text and labels to the length of the longest element in its batch, so they are a uniform length. While it is possible to pad your text in the `tokenizer` function by setting `padding=True`, dynamic padding is more efficient.

Unlike other data collators, `DataCollatorWithPadding` needs to apply a different padding method to `input_values` and `labels`. You can apply a different padding method with a context manager:

```py
>>> import torch

>>> from dataclasses import dataclass, field
>>> from typing import Any, Dict, List, Optional, Union

>>> @dataclass
... class DataCollatorCTCWithPadding:

...     processor: AutoProcessor
...     padding: Union[bool, str] = True

> ...     max_length: Optional[int] = None

Reviewer: maybe for an example we can remove this and all of the following three lines since in 99% of the cases it's not used anyways

Author: That makes sense, thanks!

...     max_length_labels: Optional[int] = None
...     pad_to_multiple_of: Optional[int] = None
...     pad_to_multiple_of_labels: Optional[int] = None

...     def __call__(self, features: List[Dict[str, Union[List[int], torch.Tensor]]]) -> Dict[str, torch.Tensor]:
...         # split inputs and labels since they have to be of different lengths and need
...         # different padding methods
...         input_features = [{"input_values": feature["input_values"]} for feature in features]
...         label_features = [{"input_ids": feature["labels"]} for feature in features]

...         batch = self.processor.pad(
...             input_features,
...             padding=self.padding,

> ...             max_length=self.max_length,

Reviewer: ```suggestion
```

> ...             pad_to_multiple_of=self.pad_to_multiple_of,

Reviewer: ```suggestion
```

...             return_tensors="pt",
...         )
...         with self.processor.as_target_processor():
...             labels_batch = self.processor.pad(
...                 label_features,
...                 padding=self.padding,

> ...                 max_length=self.max_length_labels,

Reviewer: ```suggestion
```

> ...                 pad_to_multiple_of=self.pad_to_multiple_of_labels,

Reviewer: ```suggestion
```

...                 return_tensors="pt",
...             )

...         # replace padding with -100 to ignore loss correctly
...         labels = labels_batch["input_ids"].masked_fill(labels_batch.attention_mask.ne(1), -100)

...         batch["labels"] = labels

...         return batch
```

Create a batch of examples and dynamically pad them with `DataCollatorForCTCWithPadding`:

```py
>>> data_collator = DataCollatorCTCWithPadding(processor=processor, padding=True)
```

# Fine-tune with Trainer

> Load Wav2Vec2 with [`AutoModelForCTC`]. Save GPU memory by calculating the mean of the loss reduction instead of the sum, and specify the padding token:

Reviewer: ```suggestion
Load Wav2Vec2 with [`AutoModelForCTC`]. It's often better to average the CTC loss instead of using the default configuration of CTC loss summation:
```

No memory is saved by doing `"mean"` - setting the default to `"sum"` was a bad copy-paste from `fairseq` that we can't change anymore really due to backwards breaking

...     "facebook/wav2vec-base",
...     ctc_loss_reduction="mean",
...     pad_token_id=processor.tokenizer.pad_token_id,
... )
```

<Tip>

If you aren't familiar with fine-tuning a model with the [`Trainer`], take a look at the basic tutorial [here](training#finetune-with-trainer)!

</Tip>

At this point, only three steps remain:

1. Define your training hyperparameters in [`TrainingArguments`].
2. Pass the training arguments to [`Trainer`] along with the model, datasets, tokenizer, and data collator.
3. Call [`~Trainer.train`] to fine-tune your model.

```py
>>> training_args = TrainingArguments(
...     output_dir="./results",
...     group_by_length=True,
...     per_device_train_batch_size=16,
...     evaluation_strategy="steps",
...     num_train_epochs=3,
...     fp16=True,
...     gradient_checkpointing=True,
...     learning_rate=1e-4,
...     weight_decay=0.005,
...     save_total_limit=2,
... )

>>> trainer = Trainer(
...     model=model,
...     args=training_args,
...     train_dataset=timit["train"],
...     eval_dataset=timit["test"],
...     tokenizer=processor.feature_extractor,
...     data_collator=data_collator,
... )

>>> trainer.train()
```

<Tip>

> For a more in-depth example of how to fine-tune a model for automatic speech recognition, take a look at the corresponding [PyTorch notebook](https://colab.research.google.com/github/huggingface/notebooks/blob/master/examples/speech_recognition.ipynb).

Reviewer: Think this colab is a bit out-dated. Could we instead link to those two blog posts:
- https://huggingface.co/blog/fine-tune-xlsr-wav2vec2 if people want to fine-tune multi-lingual models
- https://huggingface.co/blog/fine-tune-wav2vec2-english for the English version
